src/librustc_errors/lib.rs

```diff
@@ -431,6 +446,31 @@ impl Handler {
         self.err_count.store(0, SeqCst);
     }
 
+    /// Stash a given diagnostic with the given `Span` and `StashKey` as the key for later stealing.
+    /// If the diagnostic with this `(span, key)` already exists, this will result in an ICE.
+    pub fn stash_diagnostic(&self, span: Span, key: StashKey, diag: Diagnostic) {
+        if let Some(old) = self.stashed_diagnostics.borrow_mut().insert((span, key), diag) {
+            // We are removing a previously stashed diagnostic which should not happen.
+            // Create a builder and drop it on the floor to get an ICE.
+            drop(DiagnosticBuilder::new_diagnostic(self, old));
```

Comment:
Maybe bug with a message saying that you stashed a diagnostic twice instead.

Reply:
I don't think I can use `bug!(...)` here because that creates a cyclic dependency between `librustc` and `librustc_errors` but I can move some of the formatting infra over to `rustc_errors` and then use it here.